ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/debug/BreakpointStorageImpl.java

```diff
@@ -0,0 +1,241 @@
+                preserve();
+            });
+
+            clearOutdatedRecords();
+        }
+    }
+
+    @Override
+    public void addAll(final List<Breakpoint> breakpoints) {
+        readAllBreakpointMarker.then(onFulfilled -> {
+            BreakpointStorageImpl.this.breakpoints.addAll(breakpoints);
+
+            preserve();
+        });
+    }
+
+    @Override
+    public void add(final Breakpoint breakpoint) {
+        readAllBreakpointMarker.then(onFulfilled -> {
+            BreakpointStorageImpl.this.breakpoints.add(breakpoint);
+
+            preserve();
+        });
+    }
+
+    @Override
+    public void delete(final Breakpoint breakpoint) {
+        readAllBreakpointMarker.then(onFulfilled -> {
+            BreakpointStorageImpl.this.breakpoints.removeIf(
+                    b -> b.getLineNumber() == breakpoint.getLineNumber() && b.getPath().equals(breakpoint.getPath())
+            );
+
+            preserve();
+        });
+    }
+
+    @Override
+    public void deleteAll(final List<Breakpoint> breakpoints) {
+        readAllBreakpointMarker.then(onFulfilled -> {
+            for (Breakpoint breakpoint : breakpoints) {
+                BreakpointStorageImpl.this.breakpoints.removeIf(
+                        b -> b.getLineNumber() == breakpoint.getLineNumber() && b.getPath().equals(breakpoint.getPath())
+                );
+            }
+
+            preserve();
+        });
+    }
+
+    @Override
+    public void clear() {
+        readAllBreakpointMarker.then(onFulfilled -> {
+            BreakpointStorageImpl.this.breakpoints.clear();
+            preserve();
+        });
+    }
+
+    @Override
+    public Promise<List<Breakpoint>> readAll() {
+        return readAllBreakpointMarker.then(onFulfilled -> breakpoints, onRejected -> Collections.emptyList());
+    }
+
+    private void preserve() {
+        if (storage == null) {
+            return;
+        }
+
+        List<BreakpointDto> breakpoints2save = new LinkedList<>();
+
+        for (Breakpoint breakpoint : breakpoints) {
+            breakpoints2save.add(dtoFactory.createDto(BreakpointDto.class)
+                                           .withLocation(dtoFactory.createDto(LocationDto.class)
+                                                                   .withTarget(breakpoint.getPath())
+                                                                   .withLineNumber(breakpoint.getLineNumber())));
+        }
+
+        storage.setItem(storageKey, dtoFactory.toJson(breakpoints2save));
+    }
+
+    private Promise<Void> getReadAllBreakpointsMarker() {
+        return promiseProvider.create(Executor.create((resolve, reject) -> addWorkspaceHandler(resolve)));
+    }
+
+    private void addWorkspaceHandler(final ResolveFunction<Void> readAllMarkerResolveFunc) {
+        eventBus.addHandler(WorkspaceReadyEvent.getType(), event -> BreakpointStorageImpl.this.onWorkspaceReady(readAllMarkerResolveFunc));
```

Comment:
When `WorkspaceReadyEvent` happened in BreakpointManagerImpl before BreakpointStorageImpl then `getAll` methods will be invoked. It won't return result until `readAllBreakpointMarker` is resolved. And it can happen only if only ``WorkspaceReadyEvent` is processed in BreakpointStorageImpl.